servicetalk-concurrent-api/src/test/java/io/servicetalk/concurrent/api/PublisherSwitchMapTest.java

```diff
@@ -0,0 +1,380 @@
+            assertThat(signal, instanceOf(SwitchFlatMapSignal.class));
+            validateSignal((SwitchFlatMapSignal<Integer>) signal, 101, false);
+            signal = signals.take();
+            assertThat(signal, instanceOf(SwitchFlatMapSignal.class));
+            validateSignal((SwitchFlatMapSignal<Integer>) signal, 102, false);
+            signal = signals.take();
+            assertThat(signal, is(completeSignal));
+        } finally {
+            executor.closeAsync().toFuture().get();
+        }
+    }
+
+    @Test
+    void nullPublisher() throws InterruptedException {
+        toSource(publisher.switchMap(toSwitchFunction(i -> i == 0 ? null : Publisher.from("foo"))))
+                .subscribe(subscriber);
+
+        subscriber.awaitSubscription().request(1);
+        testSubscription.awaitRequestN(1);
+
+        publisher.onNext(0);
+
+        testSubscription.awaitRequestN(2);
+        publisher.onNext(1);
+
+        validateSignal(subscriber.takeOnNext(), "foo", false);
+        validateTerminal(publisher, subscriber, false);
+    }
+
+    @Test
+    void onErrorCancelMappedPublisher() throws InterruptedException {
+        final TestSubscription testSubscription2 = new TestSubscription();
+        final TestPublisher<String> publisher2 = new TestPublisher.Builder<String>()
+                .disableAutoOnSubscribe().build(sub -> {
+                    sub.onSubscribe(testSubscription2);
+                    return sub;
+                });
+
+        toSource(publisher.switchMap(toSwitchFunction(i -> publisher2)))
+                .subscribe(subscriber);
+
+        subscriber.awaitSubscription().request(1);
+        testSubscription.awaitRequestN(1);
+        publisher.onNext(0);
+
+        testSubscription2.awaitRequestN(1);
+        publisher2.onNext("foo");
+        validateSignal(subscriber.takeOnNext(), "foo", false);
+
+        publisher.onError(DELIBERATE_EXCEPTION);
+        testSubscription2.awaitCancelled();
+        assertThat(subscriber.awaitOnError(), is(DELIBERATE_EXCEPTION));
+    }
+
+    private static <R> void validateSignal(@Nullable SwitchFlatMapSignal<R> signal, R data, boolean isSwitched) {
+        assertThat(signal, notNullValue());
+        assertThat(signal.isSwitched(), equalTo(isSwitched));
+        assertThat(signal.data(), equalTo(data));
+    }
+
+    private static <P, S> void validateTerminal(TestPublisher<P> publisher, TestPublisherSubscriber<S> subscriber,
+                                                boolean onError) {
+        if (onError) {
+            publisher.onError(DELIBERATE_EXCEPTION);
+            assertThat(subscriber.awaitOnError(), is(DELIBERATE_EXCEPTION));
+        } else {
+            publisher.onComplete();
+            subscriber.awaitOnComplete();
+        }
+    }
+
+    /**
+     * A signal containing the data from a series of {@link Publisher}s connected in a serial fashion.
+     * @param <T> Type of the data.
+     */
+    interface SwitchFlatMapSignal<T> {
```

Comment:
This has shown to be useful but will keep internal to test for now.

When a switch happens application logic may need to take special action (e.g. replay initial/aggregated state as first event)